File: sdk-typescript/src/services/workspace.ts
/**
 * Workspace service for managing workspaces and file operations via gRPC
 */

import * as grpc from '@grpc/grpc-js';
import { Workspace, CreateWorkspaceParams, FileInfo, StorageType } from '../types';
import { WorkspaceServiceClient, createMetadata, promisifyUnary } from '../grpc';
import { convertGrpcError, isNotFound } from '../errors';

/**
 * Service for managing workspaces and file operations
 */
export class WorkspaceService {
  constructor(
    private readonly client: WorkspaceServiceClient,
    private readonly apiKey?: string
  ) {}

  private metadata(): grpc.Metadata {
    return createMetadata(this.apiKey);
  }

  // ==================== Workspace CRUD ====================

  /**
   * Create a new workspace
   */
  async create(params: CreateWorkspaceParams = {}): Promise<Workspace> {
    try {
      const request: any = {
        name: params.name,
        metadata: params.metadata || {},
      };
      if (params.storageType) {
        request.storageType = params.storageType;
      }
      const response = await promisifyUnary(
        this.client,
        this.client.createWorkspace,
        request,
        this.metadata()
      );
      return this.transformWorkspace(response.workspace);
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Get a workspace by ID
   */
  async get(id: string): Promise<Workspace> {
    try {
      const response = await promisifyUnary(
        this.client,
        this.client.getWorkspace,
        { id },
        this.metadata()
      );
      return this.transformWorkspace(response.workspace);
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * List all workspaces
   */
  async list(): Promise<Workspace[]> {
    try {
      const response = await promisifyUnary(
        this.client,
        this.client.listWorkspaces,
        {},
        this.metadata()
      );
      return (response.workspaces || []).map((w: any) => this.transformWorkspace(w));
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Delete a workspace
   */
  async delete(id: string): Promise<void> {
    try {
      await promisifyUnary(
        this.client,
        this.client.deleteWorkspace,
        { id },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  // ==================== File Operations ====================

  /**
   * Read a file from workspace
   */
  async readFile(workspaceId: string, path: string): Promise<string> {
    try {
      const response = await promisifyUnary(
        this.client,
        this.client.readFile,
        { workspaceId, path },
        this.metadata()
      );
      // Content is returned as bytes, convert to string
      const content = response.content;
      if (content instanceof Buffer) {
        return content.toString('utf-8');
      }
      return content?.toString() || '';
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Read a file as bytes from workspace
   */
  async readFileBytes(workspaceId: string, path: string): Promise<Uint8Array> {
    try {
      const response = await promisifyUnary(
        this.client,
        this.client.readFile,
        { workspaceId, path },
        this.metadata()
      );
      const content = response.content;
      if (content instanceof Buffer) {
        return new Uint8Array(content);
      }
      return new Uint8Array(content || []);
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Write a file to workspace
   */
  async writeFile(workspaceId: string, path: string, content: string | Uint8Array): Promise<void> {
    try {
      const contentBuffer = typeof content === 'string'
        ? Buffer.from(content, 'utf-8')
        : Buffer.from(content);
      await promisifyUnary(
        this.client,
        this.client.writeFile,
        { workspaceId, path, content: contentBuffer },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Create a directory in workspace
   */
  async mkdir(workspaceId: string, path: string): Promise<void> {
    try {
      await promisifyUnary(
        this.client,
        this.client.mkdir,
        { workspaceId, path },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * List directory contents in workspace
   */
  async listFiles(workspaceId: string, path: string): Promise<FileInfo[]> {
    try {
      const response = await promisifyUnary(
        this.client,
        this.client.listFiles,
        { workspaceId, path },
        this.metadata()
      );
      return (response.files || []).map((f: any) => this.transformFileInfo(f));
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Delete a file or directory in workspace
   */
  async deleteFile(workspaceId: string, path: string, recursive: boolean = false): Promise<void> {
    try {
      await promisifyUnary(
        this.client,
        this.client.deleteFile,
        { workspaceId, path, recursive },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Move/rename a file or directory in workspace
   */
  async moveFile(workspaceId: string, source: string, destination: string): Promise<void> {
    try {
      await promisifyUnary(
        this.client,
        this.client.moveFile,
        { workspaceId, source, destination },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Copy a file or directory in workspace
   */
  async copyFile(workspaceId: string, source: string, destination: string): Promise<void> {
    try {
      await promisifyUnary(
        this.client,
        this.client.copyFile,
        { workspaceId, source, destination },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Get file information in workspace
   */
  async getFileInfo(workspaceId: string, path: string): Promise<FileInfo> {
    try {
      const response = await promisifyUnary(
        this.client,
        this.client.getFileInfo,
        { workspaceId, path },
        this.metadata()
      );
      return this.transformFileInfo(response.file);
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Check if a file or directory exists in workspace
   */
  async fileExists(workspaceId: string, path: string): Promise<boolean> {
    try {
      await this.getFileInfo(workspaceId, path);
      return true;
    } catch (error) {
      if (isNotFound(error)) {
        return false;
      }
      throw error;
    }
  }

  /**
   * Check if a workspace exists
   */
  async exists(id: string): Promise<boolean> {
    try {
      await this.get(id);
      return true;
    } catch (error) {
      if (isNotFound(error)) {
        return false;
      }
      throw error;
    }
  }

  // ==================== NFS Transport ====================

  /**
   * Switch a remote workspace from gRPC to NFS transport
   */
  async registerNfsTransport(workspaceId: string, nfsUrl: string): Promise<void> {
    try {
      await promisifyUnary(
        this.client,
        this.client.registerNfsTransport,
        { workspaceId, nfsUrl },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Switch a remote workspace from NFS back to gRPC transport
   */
  async unregisterNfsTransport(workspaceId: string): Promise<void> {
    try {
      await promisifyUnary(
        this.client,
        this.client.unregisterNfsTransport,
        { workspaceId },
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  // ==================== Transform Helpers ====================

  /**
   * Transform proto Workspace to SDK Workspace type
   */
  private transformWorkspace(data: any): Workspace {
    const storageType = (data.storageType || 'managed') as StorageType;
    return {
      id: data.id,
      name: data.name || undefined,
      nfsUrl: data.nfsUrl || undefined,
      metadata: data.metadata || {},
      createdAt: this.transformTimestamp(data.createdAt),
      updatedAt: this.transformTimestamp(data.updatedAt),
      storageType,
      storageConfig: data.storageConfig || undefined,
    };
  }

  /**
   * Transform proto FileInfo to SDK FileInfo type
   */
  private transformFileInfo(data: any): FileInfo {
    return {
      name: data.name,
      path: data.path,
      type: data.type as 'file' | 'directory' | 'symlink',
      size: parseInt(data.size || '0', 10),
      modifiedAt: data.modifiedAt ? this.transformTimestamp(data.modifiedAt) : undefined,
    };
  }

  /**
   * Transform proto Timestamp to ISO string
   */
  private transformTimestamp(ts: any): string {
    if (!ts) return new Date().toISOString();
    if (ts.seconds) {
      const ms = parseInt(ts.seconds, 10) * 1000 + Math.floor((ts.nanos || 0) / 1000000);
      return new Date(ms).toISOString();
    }
    return new Date().toISOString();
  }
}

File: sdk-python/src/workspace_sdk/services/workspace.py
"""
Workspace service for managing workspaces and file operations via gRPC
"""

from typing import Optional, List
from datetime import datetime

import grpc

from workspace_sdk.types import Workspace, CreateWorkspaceParams, FileInfo
from workspace_sdk.errors import convert_grpc_error, NotFoundError
from workspace_sdk.proto.workspace.v1 import workspace_pb2, workspace_pb2_grpc


def _create_metadata(api_key: Optional[str]) -> List[tuple]:
    """Create gRPC metadata with auth token"""
    if api_key:
        return [("authorization", f"Bearer {api_key}")]
    return []


class WorkspaceService:
    """Sync service for managing workspaces and file operations via gRPC"""

    def __init__(
        self,
        stub: workspace_pb2_grpc.WorkspaceServiceStub,
        api_key: Optional[str],
        timeout: float,
    ):
        self._stub = stub
        self._api_key = api_key
        self._timeout = timeout

    def _metadata(self) -> List[tuple]:
        return _create_metadata(self._api_key)

    # ==================== Workspace CRUD ====================

    def create(self, params: Optional[CreateWorkspaceParams] = None) -> Workspace:
        """Create a new workspace"""
        req = workspace_pb2.CreateWorkspaceRequest()
        if params:
            if params.name:
                req.name = params.name
            if params.storage_type:
                req.storage_type = params.storage_type
            if params.metadata:
                req.metadata.update(params.metadata)

        try:
            resp = self._stub.CreateWorkspace(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return self._transform_workspace(resp.workspace)
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def get(self, workspace_id: str) -> Workspace:
        """Get a workspace by ID"""
        req = workspace_pb2.GetWorkspaceRequest(id=workspace_id)
        try:
            resp = self._stub.GetWorkspace(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return self._transform_workspace(resp.workspace)
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def list(self) -> List[Workspace]:
        """List all workspaces"""
        req = workspace_pb2.ListWorkspacesRequest()
        try:
            resp = self._stub.ListWorkspaces(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return [self._transform_workspace(w) for w in resp.workspaces]
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def delete(self, workspace_id: str) -> None:
        """Delete a workspace"""
        req = workspace_pb2.DeleteWorkspaceRequest(id=workspace_id)
        try:
            self._stub.DeleteWorkspace(
                req, metadata=self._metadata(), timeout=self._timeout
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    # ==================== File Operations ====================

    def read_file(self, workspace_id: str, path: str) -> str:
        """Read a file from workspace"""
        req = workspace_pb2.ReadFileRequest(workspace_id=workspace_id, path=path)
        try:
            resp = self._stub.ReadFile(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return resp.content.decode("utf-8") if isinstance(resp.content, bytes) else resp.content
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def read_file_bytes(self, workspace_id: str, path: str) -> bytes:
        """Read a file as bytes from workspace"""
        req = workspace_pb2.ReadFileRequest(workspace_id=workspace_id, path=path)
        try:
            resp = self._stub.ReadFile(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return resp.content if isinstance(resp.content, bytes) else resp.content.encode("utf-8")
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def write_file(self, workspace_id: str, path: str, content: str | bytes) -> None:
        """Write a file to workspace"""
        if isinstance(content, str):
            content = content.encode("utf-8")
        req = workspace_pb2.WriteFileRequest(
            workspace_id=workspace_id, path=path, content=content
        )
        try:
            self._stub.WriteFile(
                req, metadata=self._metadata(), timeout=self._timeout
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def mkdir(self, workspace_id: str, path: str) -> None:
        """Create a directory in workspace"""
        req = workspace_pb2.MkdirRequest(
            workspace_id=workspace_id, path=path
        )
        try:
            self._stub.Mkdir(
                req, metadata=self._metadata(), timeout=self._timeout
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def list_files(self, workspace_id: str, path: str) -> List[FileInfo]:
        """List directory contents in workspace"""
        req = workspace_pb2.ListFilesRequest(workspace_id=workspace_id, path=path)
        try:
            resp = self._stub.ListFiles(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return [self._transform_file_info(f) for f in resp.files]
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def delete_file(self, workspace_id: str, path: str, recursive: bool = False) -> None:
        """Delete a file or directory in workspace"""
        req = workspace_pb2.DeleteFileRequest(
            workspace_id=workspace_id, path=path, recursive=recursive
        )
        try:
            self._stub.DeleteFile(
                req, metadata=self._metadata(), timeout=self._timeout
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def move_file(self, workspace_id: str, source: str, destination: str) -> None:
        """Move/rename a file or directory in workspace"""
        req = workspace_pb2.MoveFileRequest(
            workspace_id=workspace_id, source=source, destination=destination
        )
        try:
            self._stub.MoveFile(
                req, metadata=self._metadata(), timeout=self._timeout
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def copy_file(self, workspace_id: str, source: str, destination: str) -> None:
        """Copy a file or directory in workspace"""
        req = workspace_pb2.CopyFileRequest(
            workspace_id=workspace_id, source=source, destination=destination
        )
        try:
            self._stub.CopyFile(
                req, metadata=self._metadata(), timeout=self._timeout
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def get_file_info(self, workspace_id: str, path: str) -> FileInfo:
        """Get file information in workspace"""
        req = workspace_pb2.GetFileInfoRequest(
            workspace_id=workspace_id, path=path
        )
        try:
            resp = self._stub.GetFileInfo(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return self._transform_file_info(resp.file)
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def exists(self, workspace_id: str, path: str) -> bool:
        """Check if a file or directory exists in workspace"""
        try:
            self.get_file_info(workspace_id, path)
            return True
        except NotFoundError:
            return False

    # ==================== NFS Transport ====================

    def register_nfs_transport(self, workspace_id: str, nfs_url: str) -> Workspace:
        """Register NFS transport for a workspace (switch from gRPC to NFS)"""
        req = workspace_pb2.RegisterNfsTransportRequest(
            workspace_id=workspace_id, nfs_url=nfs_url
        )
        try:
            resp = self._stub.RegisterNfsTransport(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return self._transform_workspace(resp.workspace)
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def unregister_nfs_transport(self, workspace_id: str) -> Workspace:
        """Unregister NFS transport for a workspace (switch back to gRPC)"""
        req = workspace_pb2.UnregisterNfsTransportRequest(
            workspace_id=workspace_id
        )
        try:
            resp = self._stub.UnregisterNfsTransport(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return self._transform_workspace(resp.workspace)
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    # ==================== Transform Helpers ====================

    def _transform_workspace(self, ws) -> Workspace:
        """Transform proto Workspace to SDK Workspace type"""
        created_at = None
        updated_at = None
        if ws.created_at:
            created_at = ws.created_at.ToDatetime().isoformat()
        if ws.updated_at:
            updated_at = ws.updated_at.ToDatetime().isoformat()

        return Workspace(
            id=ws.id,
            name=ws.name if ws.HasField("name") else None,
            nfs_url=ws.nfs_url if ws.HasField("nfs_url") else None,
            storage_type=ws.storage_type if ws.storage_type else None,
            storage_config=ws.storage_config if ws.storage_config else None,
            metadata=dict(ws.metadata) if ws.metadata else None,
            created_at=created_at,
            updated_at=updated_at,
        )

    def _transform_file_info(self, f) -> FileInfo:
        """Transform proto FileInfo to SDK FileInfo type"""
        modified_at = None
        if f.modified_at:
            modified_at = f.modified_at.ToDatetime().isoformat()

        return FileInfo(
            name=f.name,
            path=f.path,
            type=f.type,
            size=f.size,
            modified_at=modified_at,
        )
